### ISSUE TITLE
Typescript Import function gets compiler error in Visual Studio

### ISSUE THREAD

**Issue by prsmith1** [Microsoft/TypeScript#19337]
This question was first posted here: https://stackoverflow.com/questions/46827184/typescript-import-function-gets-compiler-error/46827260?noredirect=1#46827260

 In my Typescript I am using the import function as described here:
https://github.com/Microsoft/TypeScript/issues/12933

So my code is like: 

`import("../myScriptToBeImported").then((module) => {this.dosomethingWithModule(module);             }).catch(this.doSomethingWithError.bind(this));`

I am running this as part of an MVC project in Visual Studio with Webpack. Webpack compiles the Typescript without any errors and the project runs fine, yet for the above line the following errors are always shown in Visual Studio:

- TS1128 (TS) Declaration or statement expected.
- TS2391 (TS) Function implementation is missing or not immediately following the declaration.
- TS7010 (TS) 'import', which lacks return-type annotation, implicitly has an 'any' return type.

I am using Typescript version 2.4.2. How can I get rid of these errors?

For information, my tsconfig file looks like this:

`{   "compilerOptions": {     "module": "esnext",     "moduleResolution": "node",     "noEmitOnError": true,     "strict": true,     "removeComments": false,     "sourceMap": true,     "target": "es5",     "typeRoots": [       "node_modules/@types"     ],     "lib": [ "es6", "dom" ],     "baseUrl": ".",     "experimentalDecorators": true,     "emitDecoratorMetadata": true   },   "exclude": [     "node_modules",     "wwwroot"   ],   "compileOnSave": false,   "buildOnSave": false }`

**Comment by andy-ms** [comment]
Use `typescript@next` which has support for this feature.

**Comment by mhegazy** [comment]
You should not need to install `typescript@next`. This feature was part of TS 2.4 release and `typescript@2.4.2` should have it.

From the description above, VS is not using TS 2.4. can you share your project file?

**Comment by prsmith1** [comment]
Hi mhegazy - here is my csproj file:

```
<Project Sdk="Microsoft.NET.Sdk.Web">

    <PropertyGroup Label="Globals">
        <SccProjectName>SAK</SccProjectName>
        <SccProvider>SAK</SccProvider>
        <SccAuxPath>SAK</SccAuxPath>
        <SccLocalPath>SAK</SccLocalPath>
        <TypeScriptToolsVersion>Latest</TypeScriptToolsVersion>
    </PropertyGroup>

    <PropertyGroup>
      <TargetFramework>net462</TargetFramework>
    </PropertyGroup>
  
    <PropertyGroup>
      <!-- https://stackoverflow.com/questions/39176586/how-to-disable-typescript-compilation-in-net-core-projects -->
      <!-- Makes the TypeScript compilation task a no-op -->
      <TypeScriptCompileBlocked>true</TypeScriptCompileBlocked>
    </PropertyGroup>

    <ItemGroup>
      <EmbeddedResource Remove="Logging\**" />
    </ItemGroup>

    <ItemGroup>
      <None Remove="Areas\Ordering\Scripts\PatientIntake\intakeLayout.ts" />
      <None Remove="Scripts\Common\iocContainer.ts" />
    </ItemGroup>

    <ItemGroup>
        <PackageReference Include="AutoMapper" Version="6.1.1" />
        <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="3.0.1" />
        <PackageReference Include="Microsoft.NETCore.Portable.Compatibility" Version="1.0.1" />
        <PackageReference Include="Microsoft.Extensions.Logging.Filter" Version="1.1.2" />
        <PackageReference Include="Microsoft.Extensions.Options" Version="2.0.0" />
        <PackageReference Include="Newtonsoft.Json" Version="10.0.3.0" />
        <PackageReference Include="IdentityModel" Version="2.10.0" />
        <PackageReference Include="IdentityServer4.AccessTokenValidation" Version="1.2.1" />
        <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.1.1" />
        <PackageReference Include="Microsoft.AspNet.SignalR" Version="2.2.2" />
        <PackageReference Include="Microsoft.AspNetCore" Version="2.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.Authentication.Cookies" Version="2.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.Authentication.OpenIdConnect" Version="2.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.Owin" Version="2.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.Rewrite" Version="2.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.Session" Version="2.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="2.0.0" />
        <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="2.0.0" />
        <PackageReference Include="Microsoft.IdentityModel.Tokens" Version="5.1.4" />
        <PackageReference Include="Microsoft.VisualStudio.Web.BrowserLink" Version="2.0.0" />
        <PackageReference Include="NLog" Version="5.0.0-beta09" />
        <PackageReference Include="NLog.Web" Version="4.5.0" />
        <PackageReference Include="NLog.Web.AspNetCore" Version="4.4.1" />
        <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta004" />
        <PackageReference Include="Telerik.UI.for.AspNet.Core" Version="2017.3.913" />
    </ItemGroup>
    <ItemGroup>
        <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.0.0" />
    </ItemGroup>
    <PropertyGroup>
        <CodeAnalysisRuleSet>$(SolutionDir)\Brightree.ruleset</CodeAnalysisRuleSet>
        <AssetTargetFallback>portable-net45+win8</AssetTargetFallback>
        <PostBuildEvent></PostBuildEvent>
        <UserSecretsId>a31e29fc-c241-499d-81f1-fc0e546093fe</UserSecretsId>
    </PropertyGroup>
    <ItemGroup>
        <AdditionalFiles Include="$(SolutionDir)\stylecop.json" />
    </ItemGroup>
    <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
        <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
        <TreatSpecificWarningsAsErrors />
    </PropertyGroup>
    <ItemGroup>
        <Folder Include="Areas\ARManagement\Controllers\" />
    </ItemGroup>
    <ItemGroup>
      <TypeScriptCompile Include="Areas\Ordering\Scripts\PatientIntake\intakeLayout.ts" />
      <TypeScriptCompile Include="Scripts\Common\iocContainer.ts" />
    </ItemGroup>

    <ItemGroup>
      <Content Update="web.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </Content>
    </ItemGroup>

    <ItemGroup>
      <None Update="nlog.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.Load.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.Master.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.Prod.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.QA.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.QAHot.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.StageA.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.StageB.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
      <None Update="web.StageC.config">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </None>
    </ItemGroup>

    <ProjectExtensions>
       <VisualStudio><UserProperties appsettings_1Production_1json__JSONSchema="" Properties_4launchSettings_1json__JSONSchema="" appsettings_1QA_1json__JSONSchema="" appsettings_1json__JSONSchema="" package_1json__JSONSchema="" /></VisualStudio>
    </ProjectExtensions>

</Project>
```

**Comment by mhegazy** [comment]
I am unable to reproduce this locally with this file. can you share the full project, or a sample project that shows the issue?

Also noticed that your tsconfig.json is not referenced from the project file above, is this the same project?

**Comment by prsmith1** [comment]
The tsconfig is included in the project. I'm not sure why it does not show in the proj file.

I should mention that other developers working on the exact same project do not get this error, so it must be something to do with my local environment.

**Comment by aluanhaddad** [comment]
@prsmith1 Might you have the wrong TypeScript settings configured in your .csproj?
Even with `<TypeScriptCompileBlocked>true</TypeScriptCompileBlocked>`, the project settings tools version will still affect the language features recognized by the IDE.

For example
![image](https://user-images.githubusercontent.com/5034656/32100555-3aefeaaa-bae2-11e7-8e8e-a456a6a42484.png)

Compared to
![image](https://user-images.githubusercontent.com/5034656/32100488-ea25defe-bae1-11e7-84f7-4d4031cd9b1b.png)

**Comment by prsmith1** [comment]
@aluanhaddad I have tried '2.3' and 'Use latest available' (we are using 2.4), but it didn't make any difference. Looking at my extensions, the 'TypeScript Build for Microsoft Visual Studio' and 'TypeScript for Microsoft Visual Studio' extensions are both on 15.3.10723.1 and are set to auto update, so I think they're okay too.

**Comment by aluanhaddad** [comment]
Well, it will not work with '2.3' and I was thinking that you might have an older version of the SDK installed and that 'latest' might be mapping to '<2.4'.
